File: NOVEMBER-2025/24-11-2025/Second Best Minimum Spanning Tree.py
class Solution:
    class DSU:
        def __init__(self, n):
            self.parent = list(range(n))
            self.size = [1] * n

        def find(self, x):
            if self.parent[x] != x:
                self.parent[x] = self.find(self.parent[x])
            return self.parent[x]

        def connected(self, a, b):
            return self.find(a) == self.find(b)

        def unite(self, a, b):
            a = self.find(a)
            b = self.find(b)
            if a == b:
                return
            if self.size[a] < self.size[b]:
                a, b = b, a
            self.parent[b] = a
            self.size[a] += self.size[b]

    def secondMST(self, V, edges):
        # edges: list of [u, v, w]
        
        # Sort edges by weight
        e = sorted(edges, key=lambda x: x[2])

        dsu = self.DSU(V)
        mstWeight = 0
        mstEdges = 0
        
        # store MST edges
        mstEdgesList = []

        # Build MST using Kruskal
        for u, v, w in e:
            if not dsu.connected(u, v):
                dsu.unite(u, v)
                mstWeight += w
                mstEdges += 1
                mstEdgesList.append((u, v, w))

        # If MST not possible
        if mstEdges != V - 1:
            return -1

        secondBest = float('inf')

        # Try removing each MST edge
        for remU, remV, remW in mstEdgesList:
            temp = self.DSU(V)
            curWeight = 0
            cnt = 0

            for u, v, w in e:

                # Must skip both (u == remU and v == remV) AND (u == remV and v == remU)
                if (u == remU and v == remV) or (u == remV and v == remU):
                    continue

                if not temp.connected(u, v):
                    temp.unite(u, v)
                    curWeight += w
                    cnt += 1

            # Valid spanning tree and strictly > MST weight
            if cnt == V - 1 and curWeight > mstWeight:
                secondBest = min(secondBest, curWeight)

        return -1 if secondBest == float('inf') else secondBest

File: NOVEMBER-2025/24-11-2025/Second Best Minimum Spanning Tree.cpp
class Solution {
public:
    class DSU {
        vector<int> parent, sz;
    public:
        DSU(int n) {
            parent.resize(n + 1);
            sz.resize(n + 1, 1);
            for (int i = 0; i <= n; i++) parent[i] = i;
        }

        int find(int x) {
            return parent[x] == x ? x : parent[x] = find(parent[x]);
        }

        void unite(int a, int b) {
            a = find(a);
            b = find(b);
            if (a == b) return;
            if (sz[a] < sz[b]) swap(a, b);
            parent[b] = a;
            sz[a] += sz[b];
        }

        bool connected(int a, int b) {
            return find(a) == find(b);
        }
    };

    int secondMST(int V, vector<vector<int>> &edges) {
        vector<pair<int,pair<int,int>>> e;
        for (auto &x : edges) e.push_back({x[2], {x[0], x[1]}});
        sort(e.begin(), e.end());

        DSU dsu(V);
        int mstWeight = 0, mstEdges = 0;
        vector<pair<int,int>> mstEdgesList;

        for (auto &x : e) {
            int w = x.first, u = x.second.first, v = x.second.second;
            if (!dsu.connected(u, v)) {
                dsu.unite(u, v);
                mstWeight += w;
                mstEdges++;
                mstEdgesList.push_back({u, v});
            }
        }

        if (mstEdges != V - 1) return -1;

        int secondBest = INT_MAX;

        for (auto &rem : mstEdgesList) {
            DSU temp(V);
            int curWeight = 0, cnt = 0;

            for (auto &x : e) {
                int w = x.first, u = x.second.first, v = x.second.second;
                if (u == rem.first && v == rem.second) continue;
                if (!temp.connected(u, v)) {
                    temp.unite(u, v);
                    curWeight += w;
                    cnt++;
                }
            }

            if (cnt == V - 1 && curWeight > mstWeight)
                secondBest = min(secondBest, curWeight);
        }

        return (secondBest == INT_MAX ? -1 : secondBest);
    }
};